src/main/java/com/github/tomakehurst/wiremock/http/multipart/FileUpload.java

```diff
@@ -0,0 +1,761 @@
+/*
+ * Copyright (C) 2022 Thomas Akehurst
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.github.tomakehurst.wiremock.http.multipart;
+
+import static java.lang.String.format;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PipedInputStream;
+import java.io.PipedOutputStream;
+import java.io.UnsupportedEncodingException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import org.apache.commons.fileupload.FileItem;
+import org.apache.commons.fileupload.FileItemFactory;
+import org.apache.commons.fileupload.FileItemHeaders;
+import org.apache.commons.fileupload.FileItemIterator;
+import org.apache.commons.fileupload.FileItemStream;
+import org.apache.commons.fileupload.FileUploadBase;
+import org.apache.commons.fileupload.FileUploadBase.FileSizeLimitExceededException;
+import org.apache.commons.fileupload.FileUploadBase.FileUploadIOException;
+import org.apache.commons.fileupload.FileUploadBase.IOFileUploadException;
+import org.apache.commons.fileupload.FileUploadBase.InvalidContentTypeException;
+import org.apache.commons.fileupload.FileUploadBase.SizeLimitExceededException;
+import org.apache.commons.fileupload.FileUploadException;
+import org.apache.commons.fileupload.MultipartStream;
+import org.apache.commons.fileupload.ParameterParser;
+import org.apache.commons.fileupload.RequestContext;
+import org.apache.commons.fileupload.UploadContext;
+import org.apache.commons.fileupload.util.Closeable;
+import org.apache.commons.fileupload.util.FileItemHeadersImpl;
+import org.apache.commons.fileupload.util.LimitedInputStream;
+import org.apache.commons.fileupload.util.Streams;
+import org.apache.commons.io.IOUtils;
+
+/**
+ * The implementation is largely ported from {@link org.apache.commons.fileupload.FileUpload} and
+ * {@link org.apache.commons.fileupload.FileUploadBase} to support 'jakarta.servlet' instead of
+ * 'javax.servlet'. The standard support of multipart content type by Jetty in limited to
+ * 'multipart/form-data', so 'multipart/mixed' and 'multipart/related' are not recognized and parsed
+ * properly. To preserve backward compatibility and support wider range of multipart content,
+ * re-implementing this part of the upload.
+ */
+class FileUpload {
```

Comment:
@tomakehurst this is probably the most controversial change but here is the story. `Jetty` supports `multipart/form-data` out of the box, so that is cool. However, the `fileupload` components also supports `multipart/mixed` and `multipart/related` (+ attachments). The problems with  `fileupload` are: a) no releases for the last 4 years b) tight integration with `javax.servlet`.

There are 3 routes to take here:
 - drop `multipart/mixed` and `multipart/related` support, not very pleasant for backward compatibility
 - adapt `fileupload` to deal with Jakarta Servlet (this pull request)
 - drop `fileupload` and write own multipart parsers for `multipart/mixed`, `multipart/related` and attachments (may take some time)

Appreciate your opinion on the matter, thanks!

Reply:
The other consideration is licensing. It'd be the only case where we're copying source code into the project rather than depending on libraries and I'm not sure what impact this has on our ability to present WireMock as being from a single copyright source etc.

Reply:
The license is Apache v2.0 [1], should be good on this front, and there are still chances that `fileupload` gets new release with Jakarta support [2], no hints when though ...

[1] https://www.apache.org/licenses/LICENSE-2.0
[2] https://github.com/apache/commons-fileupload/tree/master/src/main/java/org/apache/commons/fileupload2/jaksrvlt

Reply:
Another solution to this that might be worth considering is adding just the servlet API back into the dependencies.

I haven't fully tested it yet, but adding `implementation 'javax.servlet:servlet-api:2.5'` to the build file seems to fix the issue and allow you to use the fileupload library.

Reply:
Technically you are right (since the Servlet API is referenced only in some static methods if I am not mistaken), but that would be quite confusing to be fair. I may work for `wiremock` specifically but have unintended consequences for the projects using it: I would imaging classpath scanning issues fe (even if we are talking about tests only).

Reply:
It would certainly complicate things for users who've already got the servlet API in their classpath from somewhere else, and it'd be a shame to make people switch to the standalone JAR just to avoid this.

So I guess the copied version is the least worst option at the moment.

There are some things that are necessary to do in order to comply with the Apache 2.0 licence, which are summarised here:
https://tldrlegal.com/license/apache-license-2.0-(apache-2.0)
The NOTICE file we'll need to include is here: https://github.com/apache/commons-fileupload/blob/commons-fileupload-1.4/NOTICE.txt

Reply:
@tomakehurst thanks a lot, I will do that shorly (travelling next couple of days), apologies for that

Reply:
@tomakehurst `NOTICE.txt` is added, https://github.com/wiremock/wiremock/pull/1942/files#diff-b810ebe4ffda4293c79d292059b45b59de7e649e86314ffd5a1ee4eb4b1e08c8, thanks!